src/pretix/plugins/paypal/payment.py

```diff
@@ -531,3 +532,18 @@ def shred_payment_info(self, obj):
             le.data = json.dumps(d)
             le.shredded = True
             le.save(update_fields=['data', 'shredded'])
+
+    def render_invoice_text(self, order: Order, payment: OrderPayment) -> str:
+        if order.status == Order.STATUS_PAID:
+            if payment.info_data.get('id', None):
+                return '{}\r\n{}: {}\r\n{}: {}'.format(
+                    _('The payment for this invoice has already been received.'),
+                    _('PayPal payment ID'),
+                    payment.info_data['id'],
+                    _('PayPal sale ID'),
+                    payment.info_data['transactions'][0]['related_resources'][0]['sale']['id']
```

Comment:
Sorry to give this another spin. For both things, we need to be able to not crash when these resources do not exist, e.g. because of some weird circumstances an incomplete payment resource is stored. At the very least, we need to catch KeyError and IndexError here :)